src/Compilers/CSharp/Test/Semantic/Semantics/DelegateTypeTests.cs

```diff
@@ -0,0 +1,1845 @@
+
+            static object?[] getData(string expr, string? expectedType) =>
+                new object?[] { expr, expectedType };
+        }
+
+        [Theory]
+        [MemberData(nameof(GetLambdaData))]
+        [MemberData(nameof(GetAnonymousMethodData))]
+        public void AnonymousFunction_ImplicitConversion(string anonymousFunction, string? expectedType)
+        {
+            var source =
+$@"class Program
+{{
+    static void Main(string[] args)
+    {{
+        System.Delegate d = {anonymousFunction};
+        System.Console.Write(d.GetDelegateTypeName());
+    }}
+}}";
+            var comp = CreateCompilation(new[] { source, s_utils }, parseOptions: TestOptions.RegularPreview, options: TestOptions.ReleaseExe);
+            if (expectedType is null)
+            {
+                comp.VerifyDiagnostics(
+                    // (5,29): error CS8915: The delegate type could not be inferred.
+                    //         System.Delegate d = x => x;
+                    Diagnostic(ErrorCode.ERR_CannotInferDelegateType, anonymousFunction).WithLocation(5, 29));
+            }
+            else
+            {
+                CompileAndVerify(comp, expectedOutput: expectedType);
+            }
+
+            var tree = comp.SyntaxTrees[0];
+            var model = comp.GetSemanticModel(tree);
+            var expr = tree.GetRoot().DescendantNodes().OfType<AnonymousFunctionExpressionSyntax>().Single();
+            var typeInfo = model.GetTypeInfo(expr);
+            if (expectedType == null)
+            {
+                Assert.Null(typeInfo.Type);
+            }
+            else
+            {
+                Assert.Equal(expectedType, typeInfo.Type.ToTestDisplayString());
+            }
+            Assert.Equal(SpecialType.System_Delegate, typeInfo.ConvertedType!.SpecialType);
+        }
+
+        [Theory]
+        [MemberData(nameof(GetLambdaData))]
+        [MemberData(nameof(GetAnonymousMethodData))]
+        public void AnonymousFunction_ExplicitConversion(string anonymousFunction, string? expectedType)
+        {
+            var source =
+$@"class Program
+{{
+    static void Main(string[] args)
+    {{
+        object o = (System.Delegate)({anonymousFunction});
+        System.Console.Write(o.GetType().GetTypeName());
+    }}
+}}";
+            var comp = CreateCompilation(new[] { source, s_utils }, parseOptions: TestOptions.RegularPreview, options: TestOptions.ReleaseExe);
+            if (expectedType is null)
+            {
+                comp.VerifyDiagnostics(
+                    // (5,20): error CS8915: The delegate type could not be inferred.
+                    //         object o = (System.Delegate)(x => x);
+                    Diagnostic(ErrorCode.ERR_CannotInferDelegateType, $"(System.Delegate)({anonymousFunction})").WithLocation(5, 20));
+            }
+            else
+            {
+                CompileAndVerify(comp, expectedOutput: expectedType);
+            }
```

Comment:
Consider asserting TypeInfo here as well.

Reply:
Same here.

---
In reply to: [618753395](https://github.com/dotnet/roslyn/pull/52448#discussion_r618753395) [](ancestors = 618753395)